src/psyclone/psyir/nodes/scoping_node.py

```diff
@@ -0,0 +1,94 @@
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors  S. Siso, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+
+''' This module contains the ScopingNode node implementation.'''
+
+from psyclone.psyir.nodes.node import Node
+from psyclone.psyir.nodes.reference import Reference
+from psyclone.psyir.symbols import SymbolTable
+
+
+class ScopingNode(Node):
+    ''' Abstract node that has an associated Symbol Table to keep track of
+    symbols declared on its scope.
+
+    :param children: the PSyIR nodes that are children of this Schedule.
+    :type children: list of :py:class:`psyclone.psyir.nodes.Node`
+    :param parent: the parent node of this Schedule in the PSyIR.
+    :type parent: :py:class:`psyclone.psyir.nodes.Node` or NoneType
+    :param symbol_table: initialise the node with a given symbol table.
+    :type symbol_table: :py:class:`psyclone.psyir.symbols.SymbolTable` or \
+            NoneType
+
+    '''
+    # Polymorphic parameter to initialize the Symbol Table of the Schedule
+    _symbol_table_class = SymbolTable
+
+    def __init__(self, children=None, parent=None, symbol_table=None):
+        super(ScopingNode, self).__init__(self, children=children,
+                                          parent=parent)
+        if symbol_table:
+            self._symbol_table = symbol_table
+        else:
+            self._symbol_table = self._symbol_table_class(self)
+
+    def _refine_copy(self, other):
+        ''' Refine the object attributes when a shallow copy is not the most
+        appropriate operation during a call to the copy() method.
+
+        :param other: object we are copying from.
+        :type other: :py:class:`psyclone.psyir.node.Node`
+
+        :raises NotImplementedError: Schedule copy is not supported yet.
+        '''
+        super(ScopingNode, self)._refine_copy(other)
+        self._symbol_table = other.symbol_table.deep_copy()
+        # pylint: disable=protected-access
+        self._symbol_table._node = self  # Associate to self
+
+        # Update of children references to point to the equally named
+        # symbol but from the new symbol table attached to self
+        for ref in self.walk(Reference):
+            if ref.symbol in other.symbol_table.symbols:
+                ref.symbol = self._symbol_table.lookup(ref.symbol.name)
+
+    @property
+    def symbol_table(self):
+        '''
+        :returns: table containing symbol information for the Schedule.
+        :rtype: :py:class:`psyclone.psyGen.SymbolTable`
+        '''
+        return self._symbol_table
```

Comment:
Please could you add an `__all__` at the bottom of this file as per usual.

Reply:
Done, also to Schedule and Container as I see just being in the `__init__.py` `__all__` does not add all documentation properly in the reference guide. Probably should be done for all nodes.

Reply:
Thanks and yes, I agree. (It might even be in the list of rules for the reviewer but I can't remember.)